proto/api_v3/query_service.proto

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax="proto3";
+
+package jaeger.api_v3;
+
+import "opentelemetry/proto/trace/v1/trace.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+
+option go_package = "api_v3";
+option java_package = "io.jaegertracing.api_v3";
+
+// Request object to get a trace.
+message GetTraceRequest {
+  // Hex encoded 128 bit trace_id
+  string trace_id = 1;
+}
+
+// A single response chunk holds a single trace.
+// Note that a subsequent request to get a trace might return a different result
+// because spans are reported in asynchronous chunks.
+message SpansResponseChunk {
+  // OTLP resource spans
+  // see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto
+  // In case of JSON format the ids (trace_id, span_id, parent_id) are encoded in base64 even though OpenTelemetry specification
+  // mandates to use hex encoding (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+  // Base64 is chosen to keep compatibility with JSONPb codec.
+  repeated opentelemetry.proto.trace.v1.ResourceSpans resource_spans = 1;
+}
+
+// Query parameters to find traces.
+message TraceQueryParameters {
+  string service_name = 1;
+  string operation_name = 2;
+  map<string, string> attributes = 3;
+  google.protobuf.Timestamp start_time_min = 4;
+  google.protobuf.Timestamp start_time_max = 5;
+  google.protobuf.Duration duration_min = 6;
+  google.protobuf.Duration duration_max = 7;
```

Comment:
should probably be more explicit about these as well - which spans do these apply to?

Reply:
+1 there are per span, right?

Reply:
what I meant was the same question as with attributes - do we consider the whole TraceQueryParameters clause to be a single conjunction that must be satisfied on a single span?

Reply:
This is a common source of confusion for our users. Most users I know expect the duration filter to apply to the trace, not to spans within the trace.

Reply:
There is a comment which should make it clear that it is for `Span` 

> // Span min duration.

Reply:
Doesn't make sense to fix this behavior for this new API? We can claim it works as documented, but users may claim it doesn't work as expected.

Reply:
I agree that the current UI search is confusing - there is duration but it does not say whether it is per span or per trace.

The current implementation uses span duration. Span duration is well defined (a span has start and end time) and can be directly queried from the storage. 

Trace duration is more complicated. It's not clearly defined and cannot be queried from the storage. The naive implementation could take the first and the last timestamp and the time in between consider as a trace duration, but is that correct? It does not consider async and messaging scenarios. It is also a breaking change to the current behavior.

Reply:
As a user, what is most useful? Searching for the duration of individual spans, or for the duration of a trace? If you think it's span duration, I'm in having it this way.

Reply:
There won't be a clear answer, both are useful. I would certainly use span duration when querying for slow traces for a given API or DB call.

When I don't have narrow search space I would like to get the slow transactions per service. This is how the UI behaves the search traces results can be sorted per trace duration.

In most cases, the root span represents trace duration. In async messaging scenarios it's not the case but I would as well argue that trace duration in those cases is not a proper duration (e.g. is the duration result of endTime-startTime or sum of span durations). Maybe in these cases users would like to sort traces based on the end time?